python/tach/parsing/modules.py

```diff
@@ -22,12 +22,51 @@ def find_duplicate_modules(modules: list[ModuleConfig]) -> list[str]:
     return duplicate_module_paths
 
 
-def build_module_tree(source_root: Path, modules: list[ModuleConfig]) -> ModuleTree:
+def find_cycle(
+    module: ModuleConfig,
+    visited: set[str],
+    path: list[str],
+    modules: list[ModuleConfig],
+) -> list[str]:
+    if module.path in visited:
+        cycle_start_index = path.index(module.path)
+        return path[cycle_start_index:] + [module.path]
+    visited.add(module.path)
+    path.append(module.path)
+    for dependency in module.depends_on:
+        dep_module = next((mod for mod in modules if mod.path == dependency), None)
+        if dep_module:
+            cycle = find_cycle(dep_module, visited, path, modules)
+            if cycle:
+                return cycle
+    visited.remove(module.path)
+    path.pop()
+    return []
+
+
+def find_modules_with_circular_dependencies(modules: list[ModuleConfig]) -> list[str]:
+    modules_with_cycles: set[str] = set()
+    for module in modules:
+        cycle = find_cycle(module, set(), [], modules)
+        if cycle:
+            modules_with_cycles.update(cycle)
+    return list(modules_with_cycles)
```

Comment:
If I understand correctly, this collapses all discovered cycles into a single set. This might make it hard for a user to understand where the cycle(s) exist in their dependencies.

Maybe instead we could have this function return a list of all discovered cycles, and the caller can raise a custom exception which holds this data?

That way, we can catch the custom exception somewhere like `cli.py` and print a nice colorful error message with all the detected cycles.

Reply:
I created a custom error in error module.
Is it okay if I make the error module import colors?